reducer/src/main/java/com/graphicsfuzz/reducer/glslreducers/EliminateInjectionMacrosVisitor.java

```diff
@@ -16,128 +16,84 @@
-import com.graphicsfuzz.common.ast.expr.ArrayIndexExpr;
+import com.graphicsfuzz.common.ast.IAstNode;
+import com.graphicsfuzz.common.ast.expr.BinOp;
 import com.graphicsfuzz.common.ast.expr.BinaryExpr;
+import com.graphicsfuzz.common.ast.expr.ConstantExpr;
 import com.graphicsfuzz.common.ast.expr.Expr;
 import com.graphicsfuzz.common.ast.expr.FunctionCallExpr;
-import com.graphicsfuzz.common.ast.expr.MemberLookupExpr;
 import com.graphicsfuzz.common.ast.expr.ParenExpr;
-import com.graphicsfuzz.common.ast.expr.TernaryExpr;
-import com.graphicsfuzz.common.ast.expr.TypeConstructorExpr;
-import com.graphicsfuzz.common.ast.expr.UnaryExpr;
-import com.graphicsfuzz.common.ast.stmt.DoStmt;
-import com.graphicsfuzz.common.ast.stmt.ForStmt;
-import com.graphicsfuzz.common.ast.stmt.IfStmt;
-import com.graphicsfuzz.common.ast.stmt.LoopStmt;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
 import com.graphicsfuzz.common.ast.visitors.StandardVisitor;
 import com.graphicsfuzz.reducer.reductionopportunities.MacroNames;
 
 public class EliminateInjectionMacrosVisitor extends StandardVisitor {
 
   @Override
-  public void visitBinaryExpr(BinaryExpr binaryExpr) {
-    super.visitBinaryExpr(binaryExpr);
-    cleanUpMacros(binaryExpr);
-  }
-
-  @Override
-  public void visitUnaryExpr(UnaryExpr unaryExpr) {
-    super.visitUnaryExpr(unaryExpr);
-    cleanUpMacros(unaryExpr);
-  }
-
-  @Override
-  public void visitParenExpr(ParenExpr parenExpr) {
-    super.visitParenExpr(parenExpr);
-    cleanUpMacros(parenExpr);
-  }
-
-  @Override
-  public void visitFunctionCallExpr(FunctionCallExpr functionCallExpr) {
-    super.visitFunctionCallExpr(functionCallExpr);
-    cleanUpMacros(functionCallExpr);
-  }
-
-  public void visitArrayIndexExpr(ArrayIndexExpr arrayIndexExpr) {
-    super.visitArrayIndexExpr(arrayIndexExpr);
-    cleanUpMacros(arrayIndexExpr);
-  }
-
-  public void visitMemberLookupExpr(MemberLookupExpr memberLookupExpr) {
-    super.visitMemberLookupExpr(memberLookupExpr);
-    cleanUpMacros(memberLookupExpr);
-  }
-
-  public void visitTernaryExpr(TernaryExpr ternaryExpr) {
-    super.visitTernaryExpr(ternaryExpr);
-    cleanUpMacros(ternaryExpr);
-  }
-
-  @Override
-  public void visitTypeConstructorExpr(TypeConstructorExpr typeConstructorExpr) {
-    super.visitTypeConstructorExpr(typeConstructorExpr);
-    cleanUpMacros(typeConstructorExpr);
-  }
-
-  @Override
-  public void visitIfStmt(IfStmt ifStmt) {
-    super.visitIfStmt(ifStmt);
-    if (MacroNames.isDeadByConstruction(ifStmt.getCondition())
-        || MacroNames.isIfWrapperFalse(ifStmt.getCondition())
-        || MacroNames.isIfWrapperTrue(ifStmt.getCondition())) {
-      ifStmt.setCondition(ifStmt.getCondition().getChild(0));
+  protected void visitChildFromParent(IAstNode child, IAstNode parent) {
+    super.visitChildFromParent(child, parent);
+    if (child instanceof FunctionCallExpr) {
+      Expr expr = (Expr) child;
```

Comment:
Make expr final (in general try to always use final unless there's a reason why you cannot).

Change Expr to FunctionCallExpr.  The functions like MacroNames.isZero(...) will accept arguments of type FunctionCallExpr, as it is a subclass of Expr, and then you can use expr instead of (FunctionCallExpr) child below.

I also suggest that you call the variable functionCallExpr rather than expr is you change its type to FunctionCallExpr.

Reply:
Fixed